src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,216 @@
+			log.log(Level.FINEST, "After simplification: " + simplified);
+			return simplified;
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE, "Houston, we have a problem!", x);
+		}
+		return null;
+	}
+
+	private static class SimplificationVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+		private Map<Expression, Constant> knownValues;
+
+		public SimplificationVisitor() {
+			stack = new Stack<Expression>();
+			knownValues = new HashMap<Expression, Constant>();
+		}
+
+		public Expression getExpression() {
+			Expression top = stack.pop();
+			top = propogateConstants(top);
+			return top;
+		}
+
+		@Override
+		public void postVisit(Variable variable) {
+			pushAndPrint(variable);
+		}
+
+		@Override
+		public void postVisit(Constant constant) {
+			pushAndPrint(constant);
+		}
+
+		@Override
+		public void postVisit(Operation operation) {
+			if (operation.getOperator().equals(Operation.Operator.EQ)) {
+				Expression l = operation.getOperand(0);
+				Expression r = operation.getOperand(1);
+				if (r instanceof Constant) {
+					final Constant constVal = (Constant) r;
+					if (l instanceof Variable) {
+						final Variable constVar = (Variable) l;
+						knownValues.put(constVar, constVal);
+					}
+				}
+			}
+			pushAndPrint(operation);
+		}
+
+		private Expression propogateConstants(Expression expression) {
+
+			if (expression instanceof Constant) {
+				return expression;
+			}
+
+			if (expression instanceof Variable) {
+				if (knownValues.get(expression) != null) {
+					return knownValues.get(expression);
+				} else {
+					return expression;
+				}
+			}
+
+			if (expression instanceof Operation) {
+				Operation op = (Operation) expression;
+				Expression l = op.getOperand(0);
+				Expression r = op.getOperand(1);
+
+				if (onlyContainsConstants(op)) {
+					// If the operation only contains constants we should evaluate it and return the constant value
+					return evaluate(op);
+				}
+
+				// Handle variable assignment as a special case so we don't end up with val == val
+				if (op.getOperator().equals(Operation.Operator.EQ)) {
+					if (l instanceof Variable && r instanceof Constant) {
+						return op;
+					}
+				}
+
+				l = propogateConstants(l);
+				r = propogateConstants(r);
+				Operation newop = new Operation(op.getOperator(), l, r);
+				return newop;
+			}
+
+			System.out.println("We should never get to this point!");
+			return expression;
+		}
+
+		private Expression evaluate(Expression expression) {
+			if (expression instanceof Constant) return expression;
+
+			assert expression instanceof Operation;
+			Operation operation = (Operation) expression;
+			Expression l = operation.getOperand(0);
+			Expression r = operation.getOperand(1);
+
+			if (l instanceof Operation) l = evaluate(l);
+			if (r instanceof Operation) r = evaluate(r);
+
+			Operation.Operator op = operation.getOperator();
+			int rint;
+			int lint;
```

Comment:
Lines 150-180 are a bit dense. An inline comment or two could help with the mental model of the situation here :female_detective: